server/src/infra/repositories/communication.repository.ts

```diff
@@ -1,19 +1,60 @@
-import { AuthService, Callback, CommunicationEvent, ICommunicationRepository } from '@app/domain';
-import { Logger } from '@nestjs/common';
-import { OnGatewayConnection, OnGatewayDisconnect, WebSocketGateway, WebSocketServer } from '@nestjs/websockets';
+import {
+  AuthService,
+  ClientEvent,
+  ICommunicationRepository,
+  OnConnectCallback,
+  OnServerEventCallback,
+  ServerEvent,
+} from '@app/domain';
+import { ImmichLogger } from '@app/infra/logger';
+import {
+  OnGatewayConnection,
+  OnGatewayDisconnect,
+  OnGatewayInit,
+  WebSocketGateway,
+  WebSocketServer,
+} from '@nestjs/websockets';
 import { Server, Socket } from 'socket.io';
 
 @WebSocketGateway({ cors: true, path: '/api/socket.io' })
-export class CommunicationRepository implements OnGatewayConnection, OnGatewayDisconnect, ICommunicationRepository {
-  private logger = new Logger(CommunicationRepository.name);
-  private onConnectCallbacks: Callback[] = [];
+export class CommunicationRepository
+  implements OnGatewayConnection, OnGatewayDisconnect, OnGatewayInit, ICommunicationRepository
+{
+  private logger = new ImmichLogger(CommunicationRepository.name);
+  private onConnectCallbacks: OnConnectCallback[] = [];
+  private onServerEventCallbacks: Record<ServerEvent, OnServerEventCallback[]> = {
+    [ServerEvent.CONFIG_UPDATE]: [],
+  };
+
+  @WebSocketServer()
+  private server?: Server;
 
   constructor(private authService: AuthService) {}
 
-  @WebSocketServer() server?: Server;
+  afterInit(server: Server) {
+    this.logger.log('Initialized websocket server');
+
+    for (const event of Object.values(ServerEvent)) {
+      server.on(event, async () => {
+        this.logger.debug(`Server event: ${event} (receive)`);
+        const callbacks = this.onServerEventCallbacks[event];
+        for (const callback of callbacks) {
+          await callback();
+        }
```

Comment:
Do we want to parallelize this by doing something like `await Promise.all(callbacks.map(callback => callback()))`

Reply:
We could but I think this is fine. Who knows if things depend on each other or stuff happening in a specific order. This is pretty predictable and performance is negligible here. Tbh, I think there is only even a single callback ever registered.